### ISSUE TITLE
Configure postgres `search_path` via Query Editor

### ISSUE THREAD

**Issue by mcls** [metabase/metabase#1498]
I'm using metabase with a postgres database that has a lot of schema's. I saw that database schema support was already implemented via #1433 , but selecting tables from a specific schema is still hard to do. 

I would like to be able to set the postgres [search_path](http://www.postgresql.org/docs/9.4/static/ddl-schemas.html#DDL-SCHEMAS-PATH) in the raw query editor.
Either via:

* A "search_path" field to the query editor. 
* Or by allow multiple SQL statements, because prefixing a query with `SET search_path =  "..."` currently doesn't work either.

I would like to help with this, but I'm new to clojure and metabase. 
And I have a couple of questions:

* I'm not sure how this would affect other database drivers? Can we show/hide fields based on which database is used? 
* Which choice do you prefer: allowing `SET search_path` or something like an extra `search_path` field on the "report_card" for db drivers that support it?

---

PS 
I also checked out the query builder.

In the query builder it shows all tables from all schemas without the schema prefix (in the tables dropdwon). I pulled in the repository locally and was able to show the schema prefix, but when there are a lot of schema's its not very usable, because you have to scroll a lot. Changing this is probably a bit more complex, so I though setting the search_path via the query editor would be a nice starting point.

**Comment by akras14** [comment]
+1 this issue. Schema support appears to be pretty poor at best.

I connected our Postgres DB and tried to create a new question, by just selecting a table name.

**Got "We couldn't understand your question."**.

Went to SQL Editor, typed in:
```
set search_path=some_schema;
select * from activity_log;
```
**Got No results were returned by the query.**

Typed in:
```
select * from some_schema.activity_log;
```
**Got the result.**

**Comment by agilliland** [comment]
Thanks for reporting on this guys.  We are continuing to improve our support for table namespacing through schemas as much as possible.  

I'm curious to get a sense for how widespread your usage of schemas goes.  How many schemas do you have in your databases?  how many tables in each schema?  do you often join tables across schemas or are you typically just working within 1 schema at a time for queries?

We just released our 0.13.0 release yesterday which includes better support for the backend to understand schemas and make them part of the querying model ( #1433 ).  We are still working on ways to bubble this up into the UI in an elegant way.  @akras14 , this release should fix your original issue where you say "We couldn't understand your question."

**Comment by camsaul** [comment]
@agilliland he's not talking about using the query builder, he's talking about writing raw SQL queries.

In Postgres, for example, the default schema search path includes `public` -- when you write raw SQL you don't need to prefix objects with `public`. 

If you use a different schema you have to prefix every name with that schema, which gets tedious. That's why you can set the search path to a different set of schemas.

There problem here is that raw SQL queries don't execute multiple statements, so setting the `search_path` has no effect.

**Comment by agilliland** [comment]
yeah, i get that.  i'm just trying to get more information about how folks are using schemas in general.

**Comment by akras14** [comment]
In our case we are using a new schema for every customer, so a lot of different schemas. It would be nice to visualize data both on per schema basis( per customer) as well as across all schemas(for all customers or a group of customers) to get the general trends.

**Comment by agilliland** [comment]
got it.  so it sounds like in that situation you probably only ever want to be working with 1 schema at a time.  e.g. when you go to work on some data needs you probably pick the company first (the schema) and then start to describe what you are trying to get from the database?

**Comment by mcls** [comment]
Our use case is very similar. We have on schema per customer/company, but we also regularly join to the "public" schema (in postgres) which contains data shared by all customers.

So indeed we pick the customer/schema, and then run the query. Currently this means prefixing every table with the schema name. Ideally we would only have to do that for the shared schemas ("public" in our case) and have the company/schema as some sort of parameter.

**Comment by camsaul** [comment]
@mcls @akras14 you can use SQL parameters to simulate the behavior that you want:

![screen shot 2016-09-19 at 12 54 58 pm](https://cloud.githubusercontent.com/assets/1455846/18647058/1a5558b8-7e69-11e6-859b-2af8a7bcded8.png)

Note that due to some limitations of our SQL parameters you'll want to make sure to make the variable a `Number` rather than `Text` to avoid single quotes being put around the identifier (see issue #3370 for more details).

Let me know if that's a workable solution to this issue so we can go ahead and close it out.

**Comment by mcls** [comment]
Yes, thanks. The parameters are great for this :)

Op ma 19 sep. 2016 22:03 schreef Cam Saül <notifications@github.com>:

**Comment by whitemark** [comment]
All, 
With the all the respect but I do not believe the issue is closed, on similar tools, example redash, you can use two (or more) sql statements in one go, example 

set search_path=some_schema;
select * from activity_log;

It would be important to have this on metabase as well, please consider also that many DB use statements to set parameters before running a query, so it is very useful to have the capacity to use multiple sql statements in a single query

Thanks!

**Comment by camsaul** [comment]
It isn't closed @whitemark. It's still open ;)

**Comment by foobarna** [comment]
Have PSQL with PostGIS and when doing a raw SQL question with pretty print for a normalised addressed, it fails:

```
select tiger.pprint_addy(norm_address) from public.customer;
```
with
```
ERROR: function cull_null(text) does not exist Hint: No function matches the given name and argument types.
You might need to add explicit type casts. Where: PL/pgSQL function tiger.pprint_addy(tiger.norm_addy)
``` 

Which makes no sense, as in the SQL editor this works as expected.

Also tried to prepend `set search_path=public,tiger;` but this results in:
```
Select statement did not produce a ResultSet for native query
```

**Comment by ixipixi** [comment]
Related to [Support database routing based on schemas](https://github.com/metabase/metabase/issues/28636)
